### PR TITLE
Add Megascale Hangs Playbook as a how-to guide

### PR DESCRIPTION
**Pull request by melissawm**
# Description

Adds the Megascale Hangs Playbook as a how-to guide. Initial content review done by @bvandermoon 

# Tests

Documentation currently builds successfully despite warnings (as flagged in #2356)

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

File: docs/guides/megascale_hang_playbook.md [hunk 1]
```diff
@@ -0,0 +1,103 @@
+# Megascale Hang Playbook
+
+Cloud TPU Multislice environments are composed of multiple TPU slices that communicate over the Data Center Network (DCN). Multislice workloads use Megascale collectives to communicate over DCN. This guide discusses how to identify and diagnose sources of performance issues such as slowness, hangs or errors in a multi-slice job driven by Megascale.
+
+## Hangs
+
+A Megascale *hang* occurs when a worker has waited on a Megascale communication operation for a set timeout period. In this situation, you will see a Megascale `HANG_DETECTED` message in your Cloud TPU logs. Because Megascale is responsible for communicating over DCN, the hang is generally first detected and reported by Megascale. However, this does not mean the error has anything to do with Megascale. Most often, hang detection is a symptom of a problem in another part of the system (e.g. unexpected job restarts, hardware errors, executable mismatches).
```
Comment (bvandermoon) [inline docs/guides/megascale_hang_playbook.md]: Can we remove this portion? `(e.g. unexpected job restarts, hardware errors, executable mismatches)`

File: docs/guides/megascale_hang_playbook.md [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+Full error digest:
+  Potential cause: INCONSISTENT_TPU_PROGRAM
+  Potential culprit workers: <host_name>
+  TPU stats:
+    <host_name>: <pc>
+  TPU program fingerprints:
+    <host_name>: <fingerprint>
+```
+
+If multiple TPU programs are detected across the workers, then it might be an issue with JAX tracing or the XLA compiler. If you see this log, we need to inspect the HLO dumps. Send these to Google using the following [steps](https://openxla.org/xla/hlo_dumps).
```
Comment (bvandermoon) [inline docs/guides/megascale_hang_playbook.md]: Can we change this to remove the first sentence `(If multiple TPU programs...)`, and then change the next part to:

```
If you see this log, we need to inspect the HLO dumps. Use the following steps to [collect HLO dumps](https://openxla.org/xla/hlo_dumps), and then send them to Google.
```

File: docs/guides/megascale_hang_playbook.md [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+```
+Megascale detects a hang but cannot determine the root cause. Please inspect the full digest below.
+```
+
+### 4. Checkpoint File Sizes Too Small
+
+If your checkpoint file sizes are too small and frequent, there is a risk of GCS issues. We recommend using [Orbax checkpointing](https://orbax.readthedocs.io/en/latest/index.html) which writes 200MB files by default.
+
+### 5. Training code uses MXLA to synchronize for a I/O operation
```
Comment (bvandermoon) [inline docs/guides/megascale_hang_playbook.md:89]: Since we don't want to use the previous code image, can we add some example code snippets that illustrate the same point?

Comment (melissawm) [reply docs/guides/megascale_hang_playbook.md:89]: Done. Let me know if this works.

File: docs/guides/megascale_hang_playbook.md [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+In case one of the hosts is taking more than the time set in `--megascale_graph_hang_threshold` (default is 5 minutes) then MXLA timeout will expire. Depending on the setting, MXLA will either block the training or will crash and restart it. If the latter, then a series of restarts will occur until the I/O operation completes within the timeout period.
+
+For example, use
+
+```python
+jax._src.distributed.global_state.client.wait_at_barrier("<barrier_id>")
+multihost_utils.sync_global_devices(...)
+```
+
+for worker synchronization.
```
Comment (bvandermoon) [inline docs/guides/megascale_hang_playbook.md]: How about something like this instead?

As an example, this could occur when using an all-reduce to synchronize at the end of checkpoint loading. One possible solution would be to switch this collective barrier to use the JAX distributed `wait_at_barrier` API instead:

```
jax._src.distributed.global_state.client.wait_at_barrier(...)
```

**Review by bvandermoon — APPROVED** [review]
